src/foam/dao/AbstractF3FileJournal.js

```diff
@@ -0,0 +1,485 @@
+}
+      `
+    },
+    // Writer uses a factory because we want to use one writer for the lifetime of this journal object
+    {
+      class: 'Object',
+      name: 'writer',
+      javaType: 'java.io.BufferedWriter',
+      javaFactory: `
+try {
+  OutputStream os = getX().get(foam.nanos.fs.Storage.class).getOutputStream(getFilename());
+  if ( os == null ) {
+    getLogger().error("File not found - journal: " + getFilename());
+  }
+  return (os == null) ? null : new BufferedWriter(new OutputStreamWriter(os));
+} catch ( Throwable t ) {
+  getLogger().error("Failed to read from journal: " + getFilename(), t);
+  throw new RuntimeException(t);
+}
+      `
+    }
+  ],
+
+  methods: [
+    {
+      name: 'put',
+      type: 'FObject',
+      args: [
+        { name: 'x',      type: 'Context' },
+        { name: 'prefix', type: 'String' },
+        { name: 'dao',    type: 'DAO' },
+        { name: 'obj',    type: 'foam.core.FObject' }
+      ],
+      javaCode: `
+        final Object id = obj.getProperty("id");
+
+        getLine().enqueue(new foam.util.concurrent.AbstractAssembly() {
+          FObject old;
+          String  record_ = null;
+
+          public Object[] requestLocks() {
+            return new Object[] { id };
+          }
+
+          public void executeUnderLock() {
+            old = dao.find_(x, id);
+            dao.put_(x, obj);
+          }
+
+          public void executeJob() {}
+
+          public void endJob() {
+            try {
+              record_ = ( old != null ) ?
+                getOutputter().stringifyDelta(old, obj) :
+                getOutputter().stringify(obj);
+            } catch (Throwable t) {
+              getLogger().error("Failed to write put entry to journal", t);
+              record_ = null;
+            }
+
+            if ( foam.util.SafetyUtil.isEmpty(record_) ) return;
+
+            try {
+              writeComment_(x, obj);
+              writePut_(
+                x,
+                record_,
+                getMultiLineOutput() ? "\\n" : "",
+                foam.util.SafetyUtil.isEmpty(prefix) ? "" : prefix + ".");
+
+                if ( isLast() ) getWriter().flush();
+            } catch (Throwable t) {
+              getLogger().error("Failed to write put entry to journal", t);
+            }
+          }
+        });
+
+        return obj;
+      `
+    },
+    {
+      name: 'writePut_',
+      javaThrows: [
+        'java.io.IOException'
+      ],
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'record',
+          type: 'String'
+        },
+        {
+          name: 'c',
+          type: 'String'
+        },
+        {
+          name: 'prefix',
+          type: 'String'
+        }
+      ],
+      javaCode: `
+        write_(sb.get()
+          .append(prefix)
+          .append("p(")
+          .append(record)
+          .append(")")
+          .append(c)
+          .toString());
```

Comment:
See comment below about making write_() take a CharSequence then you don't need this .toString().